src/covid19sim/frozen/clustering/naive.py

```diff
@@ -0,0 +1,370 @@
+        """Creates and returns a new cluster based on a single encounter message."""
+        return NaiveCluster(
+            # app-visible stuff below
+            uid=message.uid,
+            risk_level=message.risk_level
+                if isinstance(message, mu.EncounterMessage) else message.new_risk_level,
+            first_update_time=message.encounter_time,
+            latest_update_time=message.encounter_time
+                if isinstance(message, mu.EncounterMessage) else message.update_time,
+            messages=[message] if isinstance(message, mu.EncounterMessage)
+                else mu.create_encounter_from_update_message(message),
+            # debug-only stuff below
+            _real_encounter_uids=[message._sender_uid],
+            _real_encounter_times=[message._real_encounter_time],
+            _unclustered_messages=[message],  # once added, messages here should never be removed
+        )
+
+    def _get_encounter_match_score(
+            self,
+            message: mu.EncounterMessage,
+            ticks_per_uid_roll: np.int64 = 24 * 60 * 60,  # one tick per second, one roll per day
+    ) -> typing.Tuple[int, typing.Optional[int]]:
+        """Returns the match score between the provided encounter message and this cluster.
+
+        A negative return value means the message & cluster cannot correspond to the same person. A
+        zero value means that no match can be established with certainty (e.g. because the message
+        or cluster is too old). A positive value indicates a (possibly partial) match. A high
+        positive value means the match is more likely. If the returned value is equal to
+        `message_uid_bit_count`, the match is perfect.
+
+        Returns:
+            A tuple of the best match score (an integer in `[-1,message_uid_bit_count]`) and of
+            the matched encounter index (for internal use).
+        """
+        if message.risk_level != self.risk_level:
+            # if the cluster's risk level differs from the given encounter's risk level, there is
+            # no way to match the two, as the cluster should be updated first to that risk level,
+            # or we are looking at different users entirely
+            return -1, None
+        if message.encounter_time in self.uids and self.uids[message.encounter_time] != message.uid:
+            # if we already have a stored uid at the new encounter's timestamp with a different
+            # uid, there's no way this message can be merged into this cluster
+            return -1, None
+        best_match = (-1, None)
+        assert len(self.messages), "... a cluster cannot exist without past encounters?"
+        for encounter_idx, old_encounter in enumerate(self.messages):
+            match_score = mu.find_encounter_match_score(
+                # TODO: what happens if the received message is actually late, and we have an
+                #       'old message' that is more recent? (match scorer will throw)
```

Comment:
And have better exception handling. Basically, we want to ensure the clustering simply degrades if shit is whack, not breaks.

Reply:
degrades and logs stuff.